tests/unit/mixins/confirmation-test.js

```diff
@@ -1,18 +1,334 @@
+    beforeEach(function() {
+      addEventListenerStub = sandbox.stub(window, 'addEventListener');
+      removeEventListenerStub = sandbox.stub(window, 'removeEventListener');
+    });
+
+    describe('activate', function() {
+      it('adds a listener for beforeunload event', function() {
+        const subject = defaultSubject;
+        subject.activate();
+        expect(addEventListenerStub.getCall(0).args).to.deep.equal(['beforeunload', subject, false]);
+      });
+
+      it('adds a listener for unload event', function() {
+        const subject = defaultSubject;
+        subject.activate();
+        expect(addEventListenerStub.getCall(1).args).to.deep.equal(['unload', subject, false]);
+      });
+    });
+
+    describe('deactivate', function() {
+      it('removes the beforeunload listener', function() {
+        const subject = defaultSubject;
+        subject.deactivate();
+        expect(removeEventListenerStub.getCall(0).args).to.deep.equal(['beforeunload', subject, false]);
+      });
+
+      it('removes the unload listener', function() {
+        const subject = defaultSubject;
+        subject.deactivate();
+        expect(removeEventListenerStub.getCall(1).args).to.deep.equal(['unload', subject, false]);
+      });
+    });
+
+    describe('willTransition', function() {
+      let windowConfirmStub;
+      beforeEach(function() {
+        windowConfirmStub = sandbox.stub(window, 'confirm');
+      });
+
+      context('dirty transition allowed', function() {
+        let subject;
+        beforeEach(function() {
+          subject = defaultSubject;
+          sandbox.stub(subject, 'allowDirtyTransition').returns(true);
+        });
+
+        it('does not check isPageDirty', function() {
+          const isPageDirtyStub = sandbox.stub(subject, 'isPageDirty');
+
+          subject.send('willTransition');
+
+          expect(isPageDirtyStub.called).to.be.false;
+        });
+
+        it('does not confirm with the user even if page is dirty', function() {
+          subject.send('willTransition');
+
+          expect(windowConfirmStub.called).to.be.false;
+        });
+      });
+
+      context('dirty transition not allowed', function() {
+        let subject;
+        beforeEach(function() {
+          subject = defaultSubject;
+          sandbox.stub(subject, 'allowDirtyTransition').returns(false);
+          sandbox.stub(subject, 'modelFor').returns(null);
+        });
+
+        context('page is not dirty', function() {
+          beforeEach(function() {
+            sandbox.stub(subject, 'isPageDirty').returns(false);
+          });
+
+          it('does not confirm with the user', function() {
+            subject.send('willTransition');
+
+            expect(windowConfirmStub.called).to.be.false;
+          });
+
+          it('does not bubble the willTransition event', function() {
+            const retVal = subject.send('willTransition');
+            expect(retVal).to.equal(false);
+          });
```

Comment:
Is this the correct behavior? It seems like this might be an existing logic bug. Seems like if the dirty transition is not allowed, but the page isn't dirty, the `willTransition` event should continue to bubble.